packages/common/src/converter/protobuf-payload-converters.ts

```diff
@@ -121,17 +121,66 @@ export class ProtobufJsonPayloadConverter extends ProtobufPayloadConverter {
       return undefined;
     }
 
-    const jsonValue = protoJsonSerializer.toProto3JSON(value);
-
-    return this.constructPayload({
-      messageTypeName: getNamespacedTypeName(value.$type),
-      message: encode(JSON.stringify(jsonValue)),
-    });
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const jsonValue = protoJsonSerializer.toProto3JSON(value);
+
+      return this.constructPayload({
+        messageTypeName: getNamespacedTypeName(value.$type),
+        message: encode(JSON.stringify(jsonValue)),
+      });
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
   }
 
   public fromPayload<T>(content: Payload): T {
-    const { messageType, data } = this.validatePayload(content);
-    return protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+    const hasBufferChanged = setBufferInGlobal();
+    try {
+      const { messageType, data } = this.validatePayload(content);
+      const res = protoJsonSerializer.fromProto3JSON(messageType, JSON.parse(decode(data))) as unknown as T;
+      replaceBuffers(res);
+      return res;
+    } finally {
+      resetBufferInGlobal(hasBufferChanged);
+    }
+  }
+}
+
+function replaceBuffers<X>(obj: X) {
+  const replaceBuffersImpl = <Y>(value: any, key: string | number, target: Y) => {
+    if (Buffer.isBuffer(value)) {
+      // Need to copy. `Buffer` manages a pool slab, internally reused when Buffer objects are GC.
+      type T = keyof typeof target;
+      target[key as T] = new Uint8Array(value) as any;
+    } else {
+      replaceBuffers(value);
+    }
+  };
+
+  if (obj != null && typeof obj === 'object') {
+    // Performance optimization for large arrays
+    if (Array.isArray(obj)) {
+      obj.forEach(replaceBuffersImpl);
+    } else {
+      for (const [key, value] of Object.entries(obj)) {
+        replaceBuffersImpl(value, key, obj);
+      }
+    }
+  }
+}
+
+function setBufferInGlobal(): boolean {
+  if (typeof globalThis.Buffer === 'undefined') {
+    globalThis.Buffer = globalThis.constructor.constructor('return globalThis.Buffer')();
```

Comment:
Not sure what's the exact cost of these eval-style calls, but we should probably avoid doing these _repeatedly_.

Please either cache the right hand side of this line to a module level const, or pass it from `overrideGlobals` under some internal object (eg. `globalThis.__temporal_outter_globals.Buffer` or something like that).

Reply:
Done